pkg/resource/deploy/step_generator.go

```diff
@@ -945,8 +945,123 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, err
 			}
 
 			sg.sames[urn] = true
-			return []Step{NewSameStep(sg.deployment, event, old, old)}, nil
+
+			// When emitting a SameStep for an untargeted resource, we must also check
+			// for dependencies of the resource that may have been both deleted and
+			// not targeted. Consider:
+			//
+			// * When a resource is deleted from a program, no resource registration
+			//   will be sent for it. Moreover, no other resource in the program can
+			//   refer to it (since it forms no part of the program source).
+			//
+			// * In the event of an untargeted update, resources that previously
+			//   referred to the now-deleted resource will be updated and the
+			//   dependencies removed. The deleted resource will be removed from the
+			//   state later in the operation.
+			//
+			// HOWEVER, in the event of a targeted update that targets _neither the
+			// deleted resource nor its dependencies_:
+			//
+			// * The dependencies will have SameSteps emitted and their old states
+			//   will be copied into the new state.
+			//
+			// * The deleted resource will not have a resource registration sent for
+			//   it. However, by virtue of not being targeted, it will (correctly) not
+			//   be deleted from the state. Thus, its old state will be copied over
+			//   before the new snapshot is written. Alas, it will therefore appear
+			//   after the resources that depend upon it in the new snapshot, which is
+			//   invalid!
+			//
+			// We therefore have a special case where we can't rely on previous steps
+			// to have copied our dependencies over for us. We address this by
+			// manually traversing the dependencies of untargeted resources with old
+			// state and ensuring that they have SameSteps emitted before we emit our
+			// own.
+			//
+			// Note:
+			//
+			// * This traversal has to be depth-first -- we need to push steps for our
+			//   dependencies before we push a step for ourselves.
+			//
+			// * "Dependencies" here includes dependencies, property dependencies, and
+			//   deleted-with relationships.
+
+			seen := make(map[resource.URN]bool)
+			getDependencySteps := func(old *resource.State) ([]Step, error) {
+				if seen[old.URN] {
+					return nil, nil
+				}
+
+				seen[old.URN] = true
+				var steps []Step
+
+				for _, dep := range old.Dependencies {
+					generatedDep := sg.hasGeneratedStep(dep)
+					if !generatedDep {
+						depOld, has := sg.deployment.Olds()[dep]
+						if !has {
+							return nil, result.BailErrorf(
+								"dependency %s of untargeted resource %s has no old state",
+								dep,
+								urn,
+							)
+						}
+
+						depStep := NewSameStep(sg.deployment, nil, depOld, depOld)
+						steps = append(steps, depStep)
```

Comment:
Should we be setting sg.sames[dep] here? Otherwise if it shows in deps and propDeps aren't we going to make two same steps?

Reply:
So I thought the `seen` map would take care of that, though there is the question of whether that works across invocations and `sg.sames` is probably a better call.